### PR TITLE
Add new track pieces to the hybrid and single rail 

### PR DESCRIPTION
**Pull request by X123M3-256**
This PR implements four new track pieces - a large corkscrew, medium half loop, and small and large zero G rolls. It also adds sprites for all of these track pieces to the single rail coaster track and also sprites for the zero G rolls to the hybrid coaster track.

These track pieces cannot be saved in a TD6 file, but there is already an error message added for that so it will at least tell the user why it failed, I hope this won't be a problem.

Some clips of these track pieces in action can be seen here:

https://gfycat.com/shimmeringrewardingclownanemonefish

https://gfycat.com/enormoushomelyislandwhistler

### CODE REVIEW

**Comment by karst** [comment]
G2 is currently failing because the new sprites for the Raptor track have not yet been included.
I have been able to make to get rid of RideConstruction.cpp's by removing a < on line 2111 and 2514. I am not certain if it'll still function the same then though.

File: src/openrct2/ride/coaster/meta/AlpineCoaster.h [hunk 1]
```diff
@@ -1,5 +1,9 @@
 /*****************************************************************************
+<<<<<<< HEAD
```
Comment (duncanspumpkin) [inline src/openrct2/ride/coaster/meta/AlpineCoaster.h:2]: Rebase mistake

File: src/openrct2-ui/windows/RideConstruction.cpp [hunk 2]
```diff
@@ -2509,7 +2509,7 @@ class RideConstructionWindow final : public Window
             { windowPos.x + widget->left, windowPos.y + widget->top }, widget->height() + 1, colours[1], 0, 0,
             _numCurrentPossibleRideConfigurations, widget->width());
 
-        for (int32_t i = 0; i < 32; i++)
+        for (int32_t i = 0; i < 64; i++)
```
Comment (duncanspumpkin) [inline src/openrct2-ui/windows/RideConstruction.cpp:2512]: You will need to use 1ULL wherever the warning message is.

File: src/openrct2/ride/RideData.cpp [hunk 3]
```diff
@@ -160,7 +160,7 @@ const CarEntry CableLiftVehicle = {
     /* .SpriteGroups[Slopes12Banked45] = */ 0, SpritePrecision::None,
     /* .SpriteGroups[Slopes25Banked67] = */ 0, SpritePrecision::None,
     /* .SpriteGroups[Slopes25Banked90] = */ 0, SpritePrecision::None,
-    /* .SpriteGroups[SlopesInlineTwists] = */ 0, SpritePrecision::None,
+    /* .SpriteGroups[Slopes25InlineTwists] = */ 0, SpritePrecision::None,
```
Comment (duncanspumpkin) [inline src/openrct2/ride/RideData.cpp:163]: Was this change on purpose

File: src/openrct2/ride/TrackData.cpp [hunk 4]
```diff
@@ -3085,6 +3273,22 @@ static constexpr uint8_t TrackPieceLengths[TrackElemType::Count] = {
...
+    80,     // TrackElemType::LeftMediumHalfLoopUp
+    80,     // TrackElemType::RightMediumHalfLoopUp
+    80,     // TrackElemType::RightMediumHalfLoopDown
+    80,     // TrackElemType::LeftMediumHalfLoopDown
+    96,     // TrackElemType::LeftZeroGRollUp
+    96,     // TrackElemType::RightZeroGRollUp
+    96,     // TrackElemType::RightZeroGRollDown
+    96,     // TrackElemType::LeftZeroGRollDown
+    165,     // TrackElemType::LeftLargeZeroGRollUp
+    165,     // TrackElemType::RightLargeZeroGRollUp
```
Comment (duncanspumpkin) [inline src/openrct2/ride/TrackData.cpp:3289]: Make the comments align

File: src/openrct2/ride/TrackData.cpp [hunk 5]
```diff
@@ -3356,6 +3560,22 @@ static constexpr track_curve_chain gTrackCurveChain[TrackElemType::Count] = {
     { 65535, 65535 },                       // TrackElemType::FlatTrack1x1B
     { 65535, 65535 },                       // TrackElemType::FlatTrack1x4C
     { 65535, 65535 },                       // TrackElemType::FlatTrack3x3
+    { RideConstructionSpecialPieceSelected | TrackElemType::RightLargeCorkscrewDown, TRACK_CURVE_NONE },
+    { RideConstructionSpecialPieceSelected | TrackElemType::LeftLargeCorkscrewDown, TRACK_CURVE_NONE },
```
Comment (duncanspumpkin) [inline src/openrct2/ride/TrackData.cpp:3564]: Add a comment for which track piece it is for

File: src/openrct2/ride/TrackData.cpp [hunk 6]
```diff
@@ -3772,6 +3992,22 @@ static constexpr track_type_t AlternativeTrackTypes[TrackElemType::Count] = {
     TrackElemType::None, // TrackElemType::FlatTrack1x1B
     TrackElemType::None, // TrackElemType::FlatTrack1x4C
     TrackElemType::None, // TrackElemType::FlatTrack3x3
+    TrackElemType::None,
```
Comment (duncanspumpkin) [inline src/openrct2/ride/TrackData.cpp:3995]: Add comment for type as above

File: src/openrct2/ride/TrackData.cpp [hunk 7]
```diff
@@ -4043,8 +4279,25 @@ static constexpr money32 TrackPricing[TrackElemType::Count] = {
...
+    229376, // TrackElemType::RightZeroGRollUp
+    229376, // TrackElemType::LeftZeroGRollDown
+    229376, // TrackElemType::RightZeroGRollDown
+    425984, // TrackElemType::LeftLargeZeroGRollUp
+    425984, // TrackElemType::RightLargeZeroGRollUp
+    425984, // TrackElemType::LeftLargeZeroGRollDown
+    425984, // TrackElemType::RightLargeZeroGRollDown
 };
 
+
```
Comment (duncanspumpkin) [inline src/openrct2/ride/TrackData.cpp:4300]: Remove extra new line

File: src/openrct2/ride/TrackData.cpp [hunk 8]
```diff
@@ -4856,8 +5141,33 @@ static constexpr uint8_t TrackSequenceElementAllowedWallEdges[TrackElemType::Cou
...
 };
 
+//    ^
+//  0   3
+//<       >
+//  1   2
+//    v
+
+
```
Comment (duncanspumpkin) [inline src/openrct2/ride/TrackData.cpp:5169]: Keep to only one new line between items

File: src/openrct2/ride/TrackData.cpp [hunk 9]
```diff
@@ -5127,6 +5437,22 @@ static constexpr uint16_t TrackFlags[TrackElemType::Count] = {
...
+    /* TrackElemType::RightZeroGRollUp                          */   TRACK_ELEM_FLAG_UP | TRACK_ELEM_FLAG_NORMAL_TO_INVERSION | TRACK_ELEM_FLAG_STARTS_AT_HALF_HEIGHT | TRACK_ELEM_FLAG_INVERSION_TO_NORMAL,
+    /* TrackElemType::RightZeroGRollDown                        */   TRACK_ELEM_FLAG_DOWN | TRACK_ELEM_FLAG_INVERSION_TO_NORMAL,
+    /* TrackElemType::LeftZeroGRollDown                         */   TRACK_ELEM_FLAG_DOWN | TRACK_ELEM_FLAG_INVERSION_TO_NORMAL,
+    /* TrackElemType::LeftLargeZeroGRollUp                           */   TRACK_ELEM_FLAG_UP | TRACK_ELEM_FLAG_NORMAL_TO_INVERSION | TRACK_ELEM_FLAG_STARTS_AT_HALF_HEIGHT | TRACK_ELEM_FLAG_INVERSION_TO_NORMAL,
+    /* TrackElemType::RightLargeZeroGRollUp                          */   TRACK_ELEM_FLAG_UP | TRACK_ELEM_FLAG_NORMAL_TO_INVERSION | TRACK_ELEM_FLAG_STARTS_AT_HALF_HEIGHT | TRACK_ELEM_FLAG_INVERSION_TO_NORMAL,
+    /* TrackElemType::RightLargeZeroGRollDown                        */   TRACK_ELEM_FLAG_DOWN | TRACK_ELEM_FLAG_INVERSION_TO_NORMAL,
+    /* TrackElemType::LeftLargeZeroGRollDown                         */   TRACK_ELEM_FLAG_DOWN | TRACK_ELEM_FLAG_INVERSION_TO_NORMAL,
```
Comment (duncanspumpkin) [inline src/openrct2/ride/TrackData.cpp:5455]: Alignment is off

File: src/openrct2/ride/TrackData.cpp [hunk 10]
```diff
@@ -5402,7 +5728,24 @@ static constexpr rct_trackdefinition TrackDefinitions[TrackElemType::Count] =
...
+    { TRACK_ZERO_G_ROLL,            TRACK_SLOPE_DOWN_25,        TRACK_SLOPE_NONE,           TRACK_BANK_NONE,        TRACK_BANK_UPSIDE_DOWN, 0                  },  // TrackElemType::RightZeroGRollDown
+    { TRACK_ZERO_G_ROLL_LARGE,            TRACK_SLOPE_NONE,           TRACK_SLOPE_UP_60,          TRACK_BANK_UPSIDE_DOWN, TRACK_BANK_NONE,        0                  },  // TrackElemType::LeftLargeZeroGRollUp
+    { TRACK_ZERO_G_ROLL_LARGE,            TRACK_SLOPE_NONE,           TRACK_SLOPE_UP_60,          TRACK_BANK_UPSIDE_DOWN, TRACK_BANK_NONE,        0                  },  // TrackElemType::RightLargeZeroGRollUp
+    { TRACK_ZERO_G_ROLL_LARGE,            TRACK_SLOPE_DOWN_60,        TRACK_SLOPE_NONE,           TRACK_BANK_NONE,        TRACK_BANK_UPSIDE_DOWN, 0                  },  // TrackElemType::LeftLargeZeroGRollDown
+    { TRACK_ZERO_G_ROLL_LARGE,            TRACK_SLOPE_DOWN_60,        TRACK_SLOPE_NONE,           TRACK_BANK_NONE,        TRACK_BANK_UPSIDE_DOWN, 0                  },  // TrackElemType::RightLargeZeroGRollDown
```
Comment (duncanspumpkin) [inline src/openrct2/ride/TrackData.cpp:5746]: Align as much as possible

File: src/openrct2/ride/TrackData.cpp [hunk 11]
```diff
@@ -5466,6 +5863,16 @@ static int32_t EvaluatorQuarterTurn3Tiles(const int16_t progress)
     return -(progress / 2) + 134;
 }
 
+static int32_t EvaluatorMediumHalfLoopUp(const int16_t progress)
+{
+    return ((static_cast<uint16_t>(-(progress - 244))) / 4) + 51;
```
Comment (duncanspumpkin) [inline src/openrct2/ride/TrackData.cpp:5868]: This cast looks mighty suspect what are you trying to achieve here

File: src/openrct2/ride/VehicleSubpositionData.cpp [hunk 12]
```diff
@@ -21435,6 +21435,2376 @@ CREATE_VEHICLE_INFO(TrackVehicleInfo_9571BC, {
...
+    {103, 18, -73, 16, 8, 3}, {103, 18, -74, 16, 8, 3}, {104, 18, -75, 16, 8, 3}, {105, 18, -76, 16, 8, 3}, {105, 18, -77, 16, 8, 3}, 
+    {106, 18, -78, 16, 8, 3}, {106, 18, -79, 16, 8, 3}, {107, 18, -80, 16, 8, 3}, {108, 17, -82, 16, 8, 3}, {108, 17, -83, 16, 8, 3}, 
+    {109, 17, -84, 16, 8, 3}, {110, 17, -85, 16, 8, 3}, {110, 17, -86, 16, 8, 3}, {111, 17, -87, 16, 8, 3}, {111, 17, -88, 16, 8, 3}, 
+    {112, 17, -89, 16, 8, 3}, {113, 17, -91, 16, 8, 3}, {113, 17, -92, 16, 8, 3}, {114, 17, -93, 16, 8, 3}, {115, 17, -94, 16, 8, 3}, 
+    {115, 17, -95, 16, 8, 0}, {116, 17, -96, 16, 8, 0}, {116, 17, -97, 16, 8, 0}, {117, 17, -98, 16, 8, 0}, {117, 17, -99, 16, 8, 0}, 
+    {118, 17, -100, 16, 8, 0}, {119, 17, -102, 16, 8, 0}, {119, 17, -103, 16, 8, 0}, {120, 16, -104, 16, 8, 0}, {120, 16, -105, 16, 8, 0}, 
+    {121, 16, -106, 16, 8, 0}, {121, 16, -107, 16, 8, 0}, {122, 16, -108, 16, 8, 0}, {123, 16, -109, 16, 8, 0}, {123, 16, -110, 16, 8, 0}, 
+    {124, 16, -111, 16, 8, 0}, {124, 16, -112, 16, 8, 0}, {125, 16, -113, 16, 8, 0}, {125, 16, -114, 16, 8, 0}, {126, 1
```
Comment (duncanspumpkin) [inline src/openrct2/ride/VehicleSubpositionData.cpp:23806]: Remove extra new lines.

Comment (duncanspumpkin) [inline src/openrct2/ride/VehicleSubpositionData.cpp:23804]: This file needs the regex applied.

File: src/openrct2/ride/TrackData.cpp [hunk 13]
```diff
@@ -556,6 +572,23 @@ static constexpr uint8_t TrackSequenceProperties[][MaxSequencesPerPiece] = {
...
     /* TrackElemType::FlatTrack1x4C    */ { TRACK_SEQUENCE_FLAG_DIRECTION_1 | TRACK_SEQUENCE_FLAG_DIRECTION_3 | TRACK_SEQUENCE_FLAG_ORIGIN, TRACK_SEQUENCE_FLAG_DIRECTION_0 | TRACK_SEQUENCE_FLAG_DIRECTION_1 | TRACK_SEQUENCE_FLAG_DIRECTION_3, TRACK_SEQUENCE_FLAG_DIRECTION_1 | TRACK_SEQUENCE_FLAG_DIRECTION_3, TRACK_SEQUENCE_FLAG_DIRECTION_1 | TRACK_SEQUENCE_FLAG_DIRECTION_2 | TRACK_SEQUENCE_FLAG_DIRECTION_3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 },
     /* TrackElemType::FlatTrack3x3     */ { TRACK_SEQUENCE_FLAG_ORIGIN, TRACK_SEQUENCE_FLAG_DIRECTION_0 | TRACK_SEQUENCE_FLAG_DIRECTION_3, TRACK_SEQUENCE_FLAG_DIRECTION_0, TRACK_SEQUENCE_FLAG_DIRECTION_0 | TRACK_SEQUENCE_FLAG_DIRECTION_1, TRACK_SEQUENCE_FLAG_DIRECTION_3, TRACK_SEQUENCE_FLAG_DIRECTION_1, TRACK_SEQUENCE_FLAG_DIRECTION_2 | TRACK_SEQUENCE_FLAG_DIRECTION_3, TRACK_SEQUENCE_FLAG_DIRECTION_1 | TRACK_SEQUENCE_FLAG_DIRECTION_2, TRACK_SEQUENCE_FLAG_DIRECTION_2, 0, 0, 0, 0, 0, 0, 0 },
+    { 0 },
```
Comment (spacek531) [inline src/openrct2/ride/TrackData.cpp:575]: Missing track type comment

File: src/openrct2/ride/VehicleSubpositionData.cpp (hunk 12 above)
Comment (X123M3-256) [reply src/openrct2/ride/VehicleSubpositionData.cpp:23804]: Which regex?

File: src/openrct2/ride/TrackData.cpp (hunk 11 above)
Comment (X123M3-256) [reply src/openrct2/ride/TrackData.cpp:5859]: I put the cast there because the large/small half loops have it, but I couldn't figure out what it was actually there for, so I think it would be fine to just remove it.

File: src/openrct2/ride/RideData.cpp (hunk 3 above)
Comment (X123M3-256) [reply src/openrct2/ride/RideData.cpp:163]: Yes, there was a typo there

File: src/openrct2/ride/VehicleSubpositionData.cpp (hunk 12 above)
Comment (duncanspumpkin) [reply src/openrct2/ride/VehicleSubpositionData.cpp:23804]: https://github.com/OpenRCT2/OpenRCT2/pull/16164#issuecomment-1191148643

@Broxzier would you be able to format this for X7? Might be an idea to add a comment to the file with the regex.

**Comment by X123M3-256** [comment]
> I have been able to make the PR build by removing a singlular < on both line 2111 and line 2514 in RideConstruction.cpp. I am not certain if it'll still function the same then though.

So you replaced the bit shift with a comparison? Of course that won't function the same, that will just break it entirely - the correct fix is what @duncanspumpkin said.

> I also wanted to suggest to make this PR update the objects version so it directly comes with the new sprites of the trains.

Those objects have already been updated. You can't update the objects and the code with one PR because they're kept in separate repositories. If I need to change those sprites it will be another PR.

**Comment by duncanspumpkin** [comment]
> > I also wanted to suggest to make this PR update the objects version so it directly comes with the new sprites of the trains.
> 
> Those objects have already been updated. You can't update the objects and the code with one PR because they're kept in separate repositories. If I need to change those sprites it will be another PR.

We need to do an object release and then in this pr you reference the new object release. That is what the suggestion is about.

**Comment by X123M3-256** [comment]
> We need to do an object release and then in this pr you reference the new object release. That is what the suggestion is about.

What do you mean by "reference the new object release"?

**Comment by duncanspumpkin** [comment]
> > We need to do an object release and then in this pr you reference the new object release. That is what the suggestion is about.
> 
> What do you mean by "reference the new object release"?

You update the two references and hashs in the cmake and vcxproj. But that will be done by a team member.

File: src/openrct2/ride/TrackData.cpp [hunk 14]
```diff
@@ -4845,8 +5129,8 @@ static constexpr uint8_t TrackSequenceElementAllowedWallEdges[TrackElemType::Cou
...
     { 0b1010, 0b1010, 0b1010,      0,      0,      0,      0,      0,      0,      0,      0,      0,      0,      0,      0,      0 }, // TrackElemType::MultiDimFlatToDown90QuarterLoop
     { 0b1010, 0b1010, 0b1010,      0,      0,      0,      0,      0,      0,      0,      0,      0,      0,      0,      0,      0 }, // TrackElemType::MultiDimInvertedUp90ToFlatQuarterLoop
-    { 0b1010,      0,      0,      0,      0,      0,      0,      0,      0,      0,      0,      0,      0,      0,      0,      0 }, // TrackElemType::RotationControlToggle
-    {      0,      0,      0,      0,      0,      0,      0,      0,      0,      0,      0,      0,      0,      0,      0,      0 }, // TrackElemType::FlatTrack1x4A
+    { 0b1010,      0,      0,      0,      0,      0,      0,      0,      0,      0,      0,      0,      0,      0,      0,      0 }, // TrackElemType::RotationControlToggle //256
+    {      0,      0,      0,      0,      0,      0,      0,      0,      0,      0,      0,      0,      0,      0,      0,      0 }, // TrackElemType::FlatTrack1x4A
```
Comment (Broxzier) [inline src/openrct2/ride/TrackData.cpp:5133]:
```suggestion
    {      0,      0,      0,      0,      0,      0,      0,      0,      0,      0,      0,      0,      0,      0,      0,      0 }, // TrackElemType::FlatTrack1x4A
```

File: src/openrct2/ride/VehicleSubpositionData.cpp (hunk 12 above)
Comment (Broxzier) [reply src/openrct2/ride/VehicleSubpositionData.cpp:23804]: Sure, I can do that this evening. The comment can link to the public gist I made, rexed is there at the bottom.

https://gist.github.com/Broxzier/03041c723d1b23d9468ff4955ce3b659?permalink_comment_id=4239992#gistcomment-4239992

File: src/openrct2/ride/VehicleEntry.h [hunk 15]
```diff
@@ -141,6 +141,7 @@ static const std::string SpriteGroupNames[] = {
     "slopes42Banked22", "slopes42Banked45",   "slopes42Banked67", "slopes42Banked90", "slopes60Banked22",
     "corkscrews",       "restraintAnimation", "curvedLiftHill",
 };
+
```
Comment (Broxzier) [inline src/openrct2/ride/VehicleEntry.h:144]: unnecessary file change.

**Comment by karst** [comment]
You can currently not switch from Large Zero-G roll to a regular zero g roll or one that switches back. They're blacked out
image

**Comment by X123M3-256** [comment]
> You can currently not switch from any upside down element to a regular or large zero g roll. They're blacked out while Enable all drawable is turned on.

I can reproduce this issue but I'm not sure what's causing it. The weird thing is that it was working for me before, but now that I changed all the literals to 64 bit, it isn't, and I can't figure out why that would break things.

**Comment by X123M3-256** [comment]
@karst I think I've found the problem - there were some more places where the code assumed there were only 32 elements. I think I've fixed it now, (or at least it's working on my end).

**Comment by karst** [comment]
That fixed it for me too!

**Comment by Broxzier** [comment]
@X123M3-256 I've formatted the tables in VehiclesSubpositionData.cpp, but cannot push them to your branch. Please check the setting to allow maintainers to make adjustments to this PR.

**Comment by X123M3-256** [comment]
> @X123M3-256 I've formatted the tables in VehiclesSubpositionData.cpp, but cannot push them to your branch. Please check the setting to allow maintainers to make adjustments to this PR.

That's odd, I have that setting checked.

**Comment by Broxzier** [comment]
Weird, I do get this error still:

```
To github.com:X123M3-256/OpenRCT2.git
 ! [remote rejected]     x123M3-256/new_track_pieces -> x123M3-256/new_track_pieces (permission denied)
error: failed to push some refs to 'github.com:X123M3-256/OpenRCT2.git'
```

I've changed the remote from https to ssh, but that didn't change anything.

**Comment by X123M3-256** [comment]
@Broxzier Maybe you could send me the formatted version of the file so I can push it? I'm not really sure why you can't push to the branch as I definitely have "Allow edits and access to secrets by maintainers" checked.

**Comment by X123M3-256** [comment]
Wait, your commit is showing up on my branch now

**Comment by Broxzier** [comment]
Yes, I managed to push it now. Not sure what the problem was exactly, but I made a new separate branch, reset that one to the same commit, then pushed it to your remote without errors.

**Comment by Gymnasiast** [comment]
@X123M3-256 I have created an objects release and updated the references for you.

@duncanspumpkin Is there anything else that needs to be done for this PR?

**Comment by duncanspumpkin** [comment]
> @X123M3-256 I have created an objects release and updated the references for you.
> 
> @duncanspumpkin Is there anything else that needs to be done for this PR?

Nothing from me. Although @spacek531 track pr should perhaps be merged first since it's been outstanding for longer.

**Comment by Gymnasiast** [comment]
@duncanspumpkin Which PR is that, specifically?

**Comment by karst** [comment]
#16164

**Comment by Gymnasiast** [comment]
I prefer doing this first. If necessary, I’ll provide help to get the other PR merged.

**Review by Gymnasiast — APPROVED** [review]
